packages/idempotency/tests/unit/idempotentDecorator.test.ts

```diff
@@ -0,0 +1,163 @@
+/**
+ * Test Function Wrapper
+ *
+ * @group unit/idempotency/all
+ */
+
+import { IdempotencyOptions } from '../../src/types/IdempotencyOptions';
+import { PersistenceLayer, IdempotencyRecord } from '../../src/persistence';
+import { idempotent } from '../../src/idempotentDecorator';
+import { IdempotencyRecordStatus, IdempotencyRecordOptions } from '../../src/types';
+import { IdempotencyItemAlreadyExistsError, IdempotencyAlreadyInProgressError, IdempotencyInconsistentStateError, IdempotencyPersistenceLayerError } from '../../src/Exceptions';
+
+const mockSaveInProgress = jest.spyOn(PersistenceLayer.prototype, 'saveInProgress').mockImplementation();
+const mockGetRecord = jest.spyOn(PersistenceLayer.prototype, 'getRecord').mockImplementation();
+
+class PersistenceLayerTestClass extends PersistenceLayer {
+  protected _deleteRecord = jest.fn();
+  protected _getRecord = jest.fn();
+  protected _putRecord = jest.fn();
+  protected _updateRecord = jest.fn();
+}
+
+const options: IdempotencyOptions = { persistenceStore: new PersistenceLayerTestClass(), dataKeywordArgument: 'testingKey' };
+const functionalityToDecorate = jest.fn();
+
+class TestingClass {
+  @idempotent(options)
+  public testing(record: Record<string, unknown>): string {
+    functionalityToDecorate(record);
+
+    return 'Hi';
+  }
+}
+
+describe('Given a class with a function to decorate', (classWithFunction = new TestingClass()) => {
+  const keyValueToBeSaved = 'thisWillBeSaved';
+  const inputRecord = { testingKey: keyValueToBeSaved, otherKey: 'thisWillNot' };
+  beforeEach(()=> jest.clearAllMocks());
+  describe('When wrapping a function with no previous executions', () => {
+    beforeEach(async () => {
+      classWithFunction.testing(inputRecord);
+    });
+
+    test('Then it will save the record to IN_PROGRESS', () => {
```

Comment:
Suggestion: I see that in some places we use the string "IN_PROGRESS" and in others "INPROGRESS". I am in favour of the former for readability, but most importantly we should stick to one.